src/tree.rs

```diff
@@ -1419,13 +1430,31 @@ impl ProblemCounts {
     }
 }
 
+/// A record for an item or tombstone in a bookmark tree.
+#[derive(Clone, Copy, Debug)]
+pub enum Record<'t> {
+    /// The item isn't mentioned at all.
+    Unmentioned,
+    /// The item exists.
+    Exists(Node<'t>),
+    /// The item doesn't exist, but is known to be deleted.
+    Deleted,
+}
+
 /// A node in a bookmark tree that knows its parent and children, and
 /// dereferences to its item.
 #[derive(Clone, Copy, Debug)]
 pub struct Node<'t>(&'t Tree, &'t TreeEntry);
 
 impl<'t> Node<'t> {
+    /// Returns the item for this node.
+    #[inline]
+    pub fn item(&self) -> &'t Item {
```

Comment:
Makes lifetimes easier—`Deref` returns a reference tied to the node's lifetime (`&self`), but tying it to the entry's lifetime (`&'t`) means we can store references to the item and its fields.